### PR TITLE
[ISSUE #5874] implement file queue for tiered storage

### LINKED ISSUE #5874: [RIP-57] implement file queue for tiered storage
link to #5754 

I implement the following file, see [RIP-57](https://github.com/apache/rocketmq/wiki/RIP-57-Tiered-storage-for-RocketMQ#adaptation-layer) for detail.

image

### PR DESCRIPTION
**Pull request by ShadowySpirits**
close #5874

### CODE REVIEW

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        if (size >= MB)
+            return formatSize(size, MB, "MB");
+        if (size >= KB)
+            return formatSize(size, KB, "KB");
+        return formatSize(size, BYTE, "Bytes");
+    }
+
+    public static String getHash(String str) {
+        try {
+            MessageDigest md = MessageDigest.getInstance("MD5");
```
[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java:89]: It's not a problem. It is used to transfer a regular file name to a random string.

[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java:89]: In scenarios where security is not considered, MD5 is faster than other common hash algorithms.

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    public static String offset2FileName(final long offset) {
+        final NumberFormat numberFormat = NumberFormat.getInstance();
+
+        numberFormat.setMinimumIntegerDigits(20);
+        numberFormat.setMaximumFractionDigits(0);
+        numberFormat.setGroupingUsed(false);
+
+        try {
+            MessageDigest md = MessageDigest.getInstance("MD5");
```
[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java:106]: ditto

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (shy-share) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java:89]: hhhhhhhhhhhhhh Does it make sense to explain to a robot?

[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/util/TieredStoreUtil.java:89]: > hhhhhhhhhhhhhh Does it make sense to explain to a robot?

I meant to explain to the reviewer. (￣▽￣)"

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+        appendPosition += TieredCommitLog.CODA_SIZE;
+    }
+
+    public ByteBuffer read(long position, int length) {
+        return readAsync(position, length).join();
+    }
+
+    public CompletableFuture<ByteBuffer> readAsync(long position, int length) {
+        if (position < 0 || length < 0) {
+            throw new TieredStoreException(TieredStoreErrorCode.ILLEGAL_PARAM, "position or length is negative");
```
Comment (aaron-ai) [inline tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java:261]: The `future` here implies that there may be an exception, could we return `future` directly?

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+
+    public abstract long getSize();
+
+    protected abstract boolean exists();
+
+    protected abstract void createFile();
+
+    protected abstract void destroyFile();
+
+    protected abstract CompletableFuture<ByteBuffer> read0(long position, int length);
```
Comment (aaron-ai) [inline tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java:525]: Add comments for the abstract method.

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+        long realSize = 0;
+        if (cause instanceof TieredStoreException && ((TieredStoreException) cause).getPosition() > 0) {
+            realSize = ((TieredStoreException) cause).getPosition();
+        }
+        if (realSize <= 0) {
+            realSize = getSize();
+        }
+        if (realSize > 0 && realSize > commitPosition) {
+            LOGGER.error("TieredFileSegment#handleCommitException: commit failed: file: {}, try to fix position: origin: {}, real: {}", getPath(), commitPosition, realSize, cause);
+            // TODO check if this diff part is uploaded to OSS
```
Comment (aaron-ai) [inline tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java:371]: OSS? tiered-storage should not be limited to the specific implementation.

Comment (aaron-ai) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java:371]: The `check` here could be an independent method or we could implement it in `#commit0`.

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredCommitLog.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import org.apache.rocketmq.common.message.MessageQueue;
+import org.apache.rocketmq.logging.org.slf4j.Logger;
+import org.apache.rocketmq.logging.org.slf4j.LoggerFactory;
+import org.apache.rocketmq.store.tiered.common.AppendResult;
+import org.apache.rocketmq.store.tiered.common.TieredMessageStoreConfig;
+import org.apache.rocketmq.store.tiered.util.MessageBufferUtil;
+import org.apache.rocketmq.store.tiered.util.TieredStoreUtil;
+
+public class TieredCommitLog {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TieredStoreUtil.TIERED_STORE_LOGGER_NAME);
```
Comment (aaron-ai) [inline tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredCommitLog.java:30]: It seems that `logger` or `log` is more recommended, see [here](https://stackoverflow.com/questions/1417190/should-a-static-final-logger-be-declared-in-upper-case#:~:text=In%20Java%2C%20static%20final%20variables,should%20be%20in%20upper%2Dcase.). Welcome opinions from different reviewers.

**Comment by aaron-ai** [comment]
Sorry for the late review.

FYI, I left a couple of comments above.

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/common/TieredStoreExecutor.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+    private static final BlockingQueue<Runnable> FETCH_DATA_THREAD_POOL_QUEUE;
+    public static final ExecutorService FETCH_DATA_EXECUTOR;
+
+    private static final BlockingQueue<Runnable> COMPACT_INDEX_FILE_THREAD_POOL_QUEUE;
+    public static final ExecutorService COMPACT_INDEX_FILE_EXECUTOR;
+
+    static {
+        DISPATCH_THREAD_POOL_QUEUE = new LinkedBlockingQueue<>(QUEUE_CAPACITY);
+        DISPATCH_EXECUTOR = new ThreadPoolExecutor(
```
[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/common/TieredStoreExecutor.java:46]: IMO, this way is more clear. We can directly see which thread pools are there.

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/common/TieredMessageStoreConfig.java [hunk 8]
```diff
@@ -17,9 +17,126 @@
...
+    // ConsumeQueue file size, default is 100M
+    private long tieredStoreConsumeQueueMaxSize = 100 * 1024 * 1024;
+    private int tieredStoreIndexFileMaxHashSlotNum = 5000000;
+    private int tieredStoreIndexFileMaxIndexNum = 5000000 * 4;
+    // index file will force rolling to next file after idle specified time, default is 3h
+    private int tieredStoreIndexFileRollingIdleInterval = 3 * 60 * 60 * 1000;
+    private String tieredMetadataServiceProvider = "org.apache.rocketmq.store.tiered.metadata.TieredMetadataManager";
+    private String tieredBackendServiceProvider = "";
+    // file reserved time, default is 72 hour
+    private int tieredStoreFileReservedTime = 72;
```
[the remark this answers is not in the document]
Comment (ShadowySpirits) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/common/TieredMessageStoreConfig.java:79]: I meant to keep the samilar name with fileReservedTime in brokerConfig.

File: tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java (hunk 4 above)
Comment (ShadowySpirits) [reply tieredstore/src/main/java/org/apache/rocketmq/store/tiered/container/TieredFileSegment.java:525]: These methods are subject to change. I will add comments in the pr that implements the fileSegment.

**Comment by ShadowySpirits** [comment]
> Sorry for the late review.
> 
> FYI, I left a couple of comments above.

Thanks for your review. I have modified according to your comment, please have a look again.

**Review by aaron-ai — APPROVED** [review]

**Comment by aaron-ai** [comment]
Looks great!  👍

leave this pull request for another day in case other reviewers wanna take a look.

**Review by lollipopjin — APPROVED** [review]
LGTM

**Pull request merged**